packages/shared/src/server/repositories/clickhouse.ts

```diff
@@ -16,6 +16,72 @@ import {
   StorageServiceFactory,
 } from "../services/StorageService";
 
+/**
+ * Custom error class for ClickHouse resource-related errors
+ * Provides actionable advice for users
+ */
+// Error type configuration map
+const ERROR_TYPE_CONFIG: Record<
+  "MEMORY_LIMIT" | "OVERCOMMIT" | "TIMEOUT",
+  {
+    discriminators: string[];
+  }
+> = {
+  MEMORY_LIMIT: {
+    discriminators: ["memory limit exceeded"],
+  },
+  OVERCOMMIT: {
+    discriminators: ["OvercommitTracker"],
+  },
+  TIMEOUT: {
+    discriminators: ["Timeout", "timeout", "timed out"],
+  },
+};
+
+const ERROR_ADVICE_MESSAGE = [
+  "Database resource limit exceeded. Please use more specific filters or a shorter time range.",
```

Comment:
"Database resource limit exceeded. Please use more specific filters or a shorter time range."
```suggestion
  "Database resource limit exceeded. Please use more specific filters or a shorter time range. We are working to scale our systems continuously.",
```